src/compiler/checker.ts

```diff
@@ -5949,6 +5949,25 @@ namespace ts {
                     }
                 }
 
+                function isEffectiveClassSymbol(symbol: Symbol) {
+                    if (!(symbol.flags & SymbolFlags.Class)) {
+                        return false;
+                    }
+                    if (isInJSFile(symbol.valueDeclaration) && !isClassLike(symbol.valueDeclaration)) {
+                        // For a symbol that isn't syntactically a `class` in a JS file we have heuristics
+                        // that detect prototype assignments that indicate the symbol is *probably* a class.
+                        // Filter out any prototype assignments for non-class symbols, i.e.
+                        //
+                        //     let A;
+                        //     A = {};
+                        //     A.prototype.b = {};
+                        const type = getTypeOfSymbol(symbol);
+                        return some(getSignaturesOfType(type, SignatureKind.Construct))

```

Comment:
Hm, thinking about it, is there a way to adjust the binder to only apply the `Class` flag if the thing is actually constructable?

Alternatively, should we still emit this as a class, but with a `private` constructor?

Reply:
Binder has no access to type information, so there's no way to be sure because of augmentations, merging globals, etc.

Reply:
What about the `private` constructor suggestion?

Reply:
I can look into it.